pkg/controllermanager/controller/credentialsbinding/reconciler.go

```diff
@@ -0,0 +1,258 @@
+		if !controllerutil.ContainsFinalizer(credentialsBinding, gardencorev1beta1.GardenerName) {
+			return reconcile.Result{}, nil
+		}
+
+		associatedShoots, err := controllerutils.DetermineShootsAssociatedTo(ctx, r.Client, credentialsBinding)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if len(associatedShoots) == 0 {
+			log.Info("No Shoots are referencing the CredentialsBinding, deletion accepted")
+
+			mayReleaseCredentials, err := r.mayReleaseCredentials(ctx, credentialsBinding)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if mayReleaseCredentials {
+				// TODO(dimityrmirchev): this code should eventually handle workload identities as a valid credential ref
+				secret := &corev1.Secret{}
+				if err := r.Client.Get(ctx, client.ObjectKey{Namespace: credentialsBinding.CredentialsRef.Namespace, Name: credentialsBinding.CredentialsRef.Name}, secret); err == nil {
+					// Remove shoot provider label and 'referred by a secret binding' label
+					hasProviderLabel, providerLabel := getProviderLabel(secret.Labels)
+					if hasProviderLabel || metav1.HasLabel(secret.ObjectMeta, v1beta1constants.LabelCredentialsBindingReference) {
+						patch := client.MergeFrom(secret.DeepCopy())
+						delete(secret.ObjectMeta.Labels, v1beta1constants.LabelCredentialsBindingReference)
+						delete(secret.ObjectMeta.Labels, providerLabel)
+						if err := r.Client.Patch(ctx, secret, patch); err != nil {
+							return reconcile.Result{}, fmt.Errorf("failed to remove referred label from Secret: %w", err)
+						}
+					}
+					// Remove finalizer from referenced secret
+					if controllerutil.ContainsFinalizer(secret, gardencorev1beta1.ExternalGardenerName) {
+						log.Info("Removing finalizer from secret", "secret", client.ObjectKeyFromObject(secret))
+						if err := controllerutils.RemoveFinalizers(ctx, r.Client, secret, gardencorev1beta1.ExternalGardenerName); err != nil {
+							return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from secret: %w", err)
+						}
+					}
+				} else if !apierrors.IsNotFound(err) {
+					return reconcile.Result{}, err
+				}
+			}
+
+			if err := r.removeLabelFromQuotas(ctx, credentialsBinding); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			// Remove finalizer from CredentialsBinding
+			if controllerutil.ContainsFinalizer(credentialsBinding, gardencorev1beta1.GardenerName) {
+				log.Info("Removing finalizer")
+				if err := controllerutils.RemoveFinalizers(ctx, r.Client, credentialsBinding, gardencorev1beta1.GardenerName); err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+				}
+			}
+
+			return reconcile.Result{}, nil
+		}
+
+		message := fmt.Sprintf("Cannot delete CredentialsBinding, because the following Shoots are still referencing it: %+v", associatedShoots)
+		r.Recorder.Event(credentialsBinding, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, message)
```

Comment:
Shouldn't this event be of type Warning? The fact that the controller is returning error with the same message speaks that this is not normal event.

Reply:
Yes, please.